### ISSUE TITLE
MultipleObjectsReturned in /srv/www/main/mainsite/aspc/auth2/backends.py

### ISSUE THREAD

**Issue by mattdahl** [aspc/mainsite#48]
Stack trace:

```
Internal Server Error: /accounts/login/
Traceback (most recent call last):
  File "/srv/www/main/env/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 132, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/srv/www/main/mainsite/aspc/auth2/views.py", line 65, in login
    user = auth.authenticate(ticket=ticket, service=service_url)
  File "/srv/www/main/env/local/lib/python2.7/site-packages/django/contrib/auth/__init__.py", line 74, in authenticate
    user = backend.authenticate(**credentials)
  File "/srv/www/main/mainsite/aspc/auth2/backends.py", line 45, in authenticate
    user_data, is_new = UserData.objects.get_or_create(user=user)
  File "/srv/www/main/env/local/lib/python2.7/site-packages/django/db/models/manager.py", line 127, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/srv/www/main/env/local/lib/python2.7/site-packages/django/db/models/query.py", line 405, in get_or_create
    return self.get(**lookup), False
  File "/srv/www/main/env/local/lib/python2.7/site-packages/django/db/models/query.py", line 338, in get
    (self.model._meta.object_name, num)
MultipleObjectsReturned: get() returned more than one UserData -- it returned 2!
```

This is curious. `UserData` objects are only ever created using `get_or_create`, so more than one object per `User` should never exist... Maybe a new user tried to log in, hanged, tried again, and there were two different threads running that eventually each created their own `UserData` object for that user? Seems far-fetched.

We could just wrap the call in a try/catch block, but that doesn't really solve the heart of the issue... There should only ever be one `UserData` object associated with a `User`! If those diverge, we will run into data consistency problems with other authentication-related stuff.

**Comment by mattdahl** [comment]
I think this is actually directly related to #46. I queried the `UserData` table for duplicate objects, but nothing was returned! I think that there is some weird race-condition on the `User` object that happens when `user.save()` triggers that recursion overflow in #46. There's not actually duplicate data floating around that we have to worry about fixing. Hopefully #47 has fixed this bug too.

**Comment by mattdahl** [comment]
Given that several more of these errors have happened in the past few days, I guess #47 didn't fix this after all.

I looked at the `UserData` table again and it turns out that there is a culprit - the `cr012014@MyMail.pomona.edu` account. It's a freshman (Chilinh Rozen) who lives in Wig. His `User` object is associated with two `UserData` objects. We don't have a `date_created` field on that table, but the ids for each object are right next to each other - 647 and 648. This leads me to believe that the duplicate object was created during the same login attempt (either during one request or over multiple, but not too far apart because no one else managed to log in during the interim). That said, this still doesn't really explain why `get_or_create()` decided to create a whole new object instead of just getting the old one.

My solution: I deleted one of Chilinh's extra `UserData` objects. This should have no side effects because we don't actually use that class for anything yet. Still don't have any idea how to fix the root of this issue, if it ever pops up again.
